src/irmin-pack/inode.ml

```diff
@@ -660,44 +669,47 @@ struct
       | Unsorted_pointers t -> Error (`Unsorted_pointers t)
 
     let hash t = Lazy.force t.hash
+    let is_node t = t.kind = Kind.Node [@@inline]
 
     let is_root t =
       match t.v with
       | Tree { depth; _ } -> depth = 0
       | Values _ ->
-          (* When [t] is of tag [Values], then [t] is root iff [t] is stable. It
-             is implied by the following.
-
-             When [t] is stable, then [t] is a root, because:
-              - Only 2 functions produce stable inodes: [stabilize] and [empty].
-              - Only the roots are output of [stabilize].
-              - An empty map can only be located at the root.
-
-             When [t] is a root of tag [Value], then [t] is stable, because:
-             - All the roots are output of [stabilize].
-             - When an unstable inode enters [stabilize], it becomes stable if
-               it has at most [Conf.stable_hash] leaves.
-             - A [Value] has at most [Conf.stable_hash] leaves because
-               [Conf.entries <= Conf.stable_hash] is enforced.
+          (* When [t] is of tag [Values], then [t] is root iff [t] is a [Node].
+             It is implied by the following;:
+
+             When [t] is a [Node], then [t] is a root, because:
+               - Only 2 functions produce stable inodes: [stabilize] and [empty].
+               - Only the roots are output of [stabilize].
+               - An empty map can only be located at the root.
+
+             When [t] is a root of tag [Value], then [t] is a stable [Inode]
+             (hashed as a [Node]), because:
+               - All the roots are output of [stabilize].
+               - When an unstable inode enters [stabilize], it becomes a
+                 stable [Inode] if it has at most [Conf.stable_hash] leaves.
+               - A [Value] has at most [Conf.stable_hash] leaves because
+                 [Conf.entries <= Conf.stable_hash] is enforced.
           *)
-          t.stable
+          is_node t
 
     let check_write_op_supported t =
       if not @@ is_root t then
         failwith "Cannot perform operation on non-root inode value."
 
     let stabilize layout t =
-      if t.stable then t
-      else
-        let n = length t in
-        if n > Conf.stable_hash then t
-        else
-          let hash =
-            lazy
-              (let vs = list layout t in
-               Node.hash (Node.v vs))
-          in
-          { hash; stable = true; v = t.v }
+      match t.kind with
+      | Node -> t
+      | Inode ->
```

Comment:
The name of `kind`'s tags are very vague in this context (i.e. `Node`/`Inode`). Even if the current magic codes are `N`/`I`, I think we shouldn't name the variants tag accordingly.

Let's rename those for something like `Node` -> `V1_stable`, `Inode` -> `V1_unstable`. This way it will be very explicit what these refer to.

Reply:
The `_stable` `_unstable` distinction is not super clear in that context either

Reply:
True.

A solution would be to isolate the `kind` type to the `Compress` module, and change `Val_impl` to instead store the `is_root` and `version` informations, instead of `kind`/`stable`, which can be computed from a function.

Reply:
From what I understand, when a node is first created (unless it's an empty node) it is implemented as an inode and viewed as one until we stabilise it. Once stabilised, if the node has less than `Conf.stable_hash` values it's not recomputed but its hash is computed as if it was a node. So the "kind" of this stabilised node is actually a `Node` or an `Inode`, no? Why do you think it's vague?

Reply:
Yes, what you describe is true. 

> it is implemented as an inode and viewed as one

Inodes are always implemented as inodes and always viewed as inodes. It is just the prehashing function that changes: in the `Node` case it is computed using `Node : Irmin.Private.Node.S` and in the `Inode` case it is computed using `Bin`.

I find the `kind = Node` / `Inode` terms poorly describe that complex situation.

Reply:
We should come to a consensus here for this PR to be merged ;-)